src/services/AbstractService.spec.ts

```diff
@@ -0,0 +1,301 @@
+			await mockService.pubEnsureMeta('0xDummyBlockHash');
+			expect(mockService['txVersion'].toNumber()).toBe(2);
+			expect(mockService['specVersion'].toNumber()).toBe(17);
+			expect(api.registry.setMetadata).not.toHaveBeenCalled();
+
+			api.registry.setMetadata = setMetadata;
+		});
+
+		it('resets metadata and correctly changes versions when versions have changed', async () => {
+			api.registry.setMetadata = jest.fn();
+
+			const mockService = await mockServiceSetup();
+			expect(api.registry.setMetadata).toHaveBeenCalledTimes(1);
+
+			api.registry.setMetadata = jest.fn();
+
+			api.rpc.state.getRuntimeVersion = (_hash: string) =>
+				Promise.resolve().then(() => {
+					return {
+						transactionVersion: new U32(kusamaRegistry, 3),
+						specVersion: new U32(kusamaRegistry, 18),
+						specName: 'kusama',
+					};
+				});
+
+			await mockService.pubEnsureMeta('0xDummyBlockHash');
+			expect(mockService['txVersion'].toNumber()).toBe(3);
+			expect(mockService['specVersion'].toNumber()).toBe(18);
+			expect(api.registry.setMetadata).toHaveBeenCalledTimes(1);
+
+			api.rpc.state.getRuntimeVersion = getRuntimeVersion;
+			api.registry.setMetadata = setMetadata;
+		});
+
+		it('does not call api.registry.register when versions have not changed', async () => {
+			api.registry.register = jest.fn();
+
+			const mockService = await mockServiceSetup();
+			expect(api.registry.register).toHaveBeenCalledTimes(1);
+			expect(api.registry.register).toBeCalledWith({
+				mock: 'specTypes!',
+			});
+
+			api.registry.register = jest.fn();
+
+			await mockService.pubEnsureMeta('0xDummyBlockHash');
+			expect(mockService['txVersion'].toNumber()).toBe(2);
+			expect(mockService['specVersion'].toNumber()).toBe(17);
+			expect(api.registry.register).not.toHaveBeenCalled();
+
+			api.registry.register = register;
+		});
+
+		it('calls api.registry.register when versions have changed', async () => {
+			api.registry.register = jest.fn();
+
+			const mockService = await mockServiceSetup();
+			expect(api.registry.register).toHaveBeenCalledTimes(1);
+			expect(api.registry.register).toBeCalledWith({
+				mock: 'specTypes!',
+			});
+
+			api.rpc.state.getRuntimeVersion = (_hash: string) =>
+				Promise.resolve().then(() => {
+					return {
+						transactionVersion: new U32(kusamaRegistry, 3),
+						specVersion: new U32(kusamaRegistry, 18),
+						specName: 'kusama',
+					};
+				});
+
+			api.registry.register = jest.fn();
+
+			await mockService.pubEnsureMeta('0xDummyBlockHash');
+			expect(mockService['txVersion'].toNumber()).toBe(3);
+			expect(mockService['specVersion'].toNumber()).toBe(18);
+			expect(api.registry.register).toHaveBeenCalledTimes(1);
+			expect(api.registry.register).toBeCalledWith({
+				mock: 'specTypes!',
+			});
+
+			api.rpc.state.getRuntimeVersion = getRuntimeVersion;
+			api.registry.register = register;
+		});
+
+		it('resets metadata and just changes `txVersion` when just the `txVersion` has changed', async () => {
```

Comment:
This _should_ never happen, and there are guards within frame-system to prevent it, so it would basically be negligence on the part of governance, but it's OK as a test.